frida_mode/src/main.c

```diff
@@ -0,0 +1,92 @@
+#include <unistd.h>
+#include <sys/personality.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+
+#include "frida-gum.h"
+#include "config.h"
+#include "debug.h"
+
+#include "interceptor.h"
+#include "instrument.h"
+#include "prefetch.h"
+
+typedef int *(*main_fn_t)(int argc, char **argv, char **envp);
+
+static main_fn_t   main_fn = NULL;
+static GumStalker *stalker = NULL;
+
+int __libc_start_main(int *(main)(int, char **, char **), int argc,
+                      char **ubp_av, void (*init)(void), void (*fini)(void),
+                      void (*rtld_fini)(void), void(*stack_end));
+
+extern void __afl_manual_init();
+
+static int on_fork() {
+
+  prefetch_read(stalker);
+  return fork();
+
+}
+
+int *on_main(int argc, char **argv, char **envp) {
+
+  GumInterceptor *interceptor = gum_interceptor_obtain();
+
+  gum_interceptor_begin_transaction(interceptor);
+  gum_interceptor_revert(interceptor, __libc_start_main);
+  gum_interceptor_end_transaction(interceptor);
+  gum_interceptor_flush(interceptor);
+
+  stalker = gum_stalker_new();
+  if (stalker == NULL) { FATAL("Failed to initialize stalker"); }
+
+  gum_stalker_set_trust_threshold(stalker, 0);
+
+  GumStalkerTransformer *transformer =
+      gum_stalker_transformer_make_from_callback(instr_basic_block, NULL, NULL);
+
+  prefetch_init();
+
+  intercept(fork, on_fork, stalker);
+
+  gum_stalker_follow_me(stalker, transformer, NULL);
+  gum_stalker_deactivate(stalker);
+
+  __afl_manual_init();
+
+  /* Child here */
+  prefetch_start(stalker);
+  return main_fn(argc, argv, envp);
+
+}
+
+int on_libc_start_main(int *(main)(int, char **, char **), int argc,
+                       char **ubp_av, void (*init)(void), void (*fini)(void),
+                       void (*rtld_fini)(void), void(*stack_end)) {
+
+  main_fn = main;
+  intercept(main, on_main, NULL);
+  return __libc_start_main(main, argc, ubp_av, init, fini, rtld_fini,
+                           stack_end);
+
+}
+
+__attribute__((constructor)) static void init() {
+
+#ifndef __APPLE__
+  (void)personality(ADDR_NO_RANDOMIZE);  // disable ASLR
+#endif
+
+  gum_init_embedded();
+  if (!gum_stalker_is_supported()) {
+
+    gum_deinit_embedded();
+    FATAL("Failed to initialize embedded");
+
+  }
+
+  intercept(__libc_start_main, on_libc_start_main, NULL);
```

Comment:
`__libc_start_main` is a glibc specific thing, so we should probably use something like `gum_darwin_find_entrypoint`. Not sure what the platform agnostic API is to find the entrypoint.